### ISSUE TITLE
How do you start reading a Tiled Editor Map exported to Lua?

### ISSUE THREAD

**Issue by LedLoaf** [ThePhD/sol2#1157]
Hello,

So I've spent 5 hours and was up until 4:30 am trying to just read the first line with Sol and an exported Tiled Editor Map to Lua.
The start of the file looks like this:

`return {
  version = "1.4",
  luaversion = "5.1",
  tiledversion = "1.4.1",
  orientation = "orthogonal",
  renderorder = "right-down",
  width = 32,
  height = 20,
  tilewidth = 16,
  tileheight = 16,
  nextlayerid = 8,
  nextobjectid = 1,
  properties = {},
  tilesets = {......}
   ...
   }`

So I believe it's returning a table and I can't seem to get the first variable even. Can someone direct me to a proper tutorial example or information, because I'm wasting so much time at this point! But I really want to get it started and I'm amazed at myself not being able to figure this out. This can't be as hard as I'm making it. There is a lot of examples and documentation, so I'll be continuing to read them and looking for answers.

Thanks! 

PS: I used an std::ifstream on the file and entered it as a big string using load_script and then executing it. I thought I could do something like lua["version"] at that point, but it didn't work. And I'd like to mention a majority of the time I'm getting an error like this **"[sol3] An error occurred and panic has been invoked: stack index -1, expected string, received nil"**

**Comment by ThePhD** [comment]
Things returned from scripts are as-if the `script`/`safe_script` call is invoking a function. In fact, as you went through the `.load` and similar rigamarole, you had to convert it to a `function` type: the result contains the returned object(s).

Here's a quick example:

script.lua
```lua
return { version = "1.4", luaversion = "5.1", tiledversion = "1.4.1", orientation = "orthogonal", renderorder = "right-down", width = 32, height = 20, tilewidth = 16, tileheight = 16 }
```

```cpp
#include <sol/sol.hpp>

int main () {
     sol::state lua;
     lua.open_libraries();

     sol::protected_function_result result = lua.safe_script_file("script.lua");
     if (result.valid()) {
          sol::error err = result;
          std::cout << err.what() << std::endl;
     }
     sol::table tile_table = result;     

     std::string luaversion = tile_table["version"];
     // etc, etc.

     return 0;
}
```

Good luck!

**Comment by LedLoaf** [comment]
Aw man, thank you very much for the help.

Unfortunately, still having problems. On the right track though, and I've been reading your "table" tutorials. I don't see any examples, where it returns {   ...  } and not just a table = {   }. This could be my lack of knowledge with Lua and understanding how to correlate it to the examples.

**"[sol3] An error occurred and panic has been invoked: stack index 1, expected table, received no value: bad get from protected_function_result (is an error)"**

I've copied your example directly, I also tried to paste it in raw to safe_script. 

I'm a little worried that I don't have this linked up correctly, yet so far all your tutorial examples have worked fine. I'm just scratching my head!

**Comment by ThePhD** [comment]
Other than a mistake in the example I gave you (using `if (result.valid())` instead of `if (!result.valid())` (an extra NOT symbol to say if it's NOT valid, the exclamation point), this works:

![](https://i.imgur.com/8PESbZT.png)

Are you sure everything is set up okay?

**Comment by LedLoaf** [comment]
Thanks ThePhD,

I quickly came back and you already pointed it out. It was saying if(NOTresult.valid()). I think I overlooked it because I've been getting errors for 5 hours, so I didn't stop and look haha. 

It worked! Much appreciated man.

**Comment by LedLoaf** [comment]
`....
nextlayerid = 8,
  nextobjectid = 1,
  properties = {},
  tilesets = {
    {
      name = "platformer",
      firstgid = 1,
      filename = "platformer.tsx",
      tilewidth = 16,
      tileheight = 16,
      spacing = 0,
      margin = 0,
      columns = 10,
      image = "Foreground/Tileset.png",
      imagewidth = 160,
      imageheight = 96,
      objectalignment = "unspecified",
      tileoffset = {
        ...
      },
      grid = {
       ...
      },
      properties = {},
      terrains = {},
      tilecount = 60,
      tiles = {}
    },`

Doing great until I hit the nested stuff.
**sol::table tilesets = tile_table.get<sol::table>("tilesets");**
These are nested containers, so I looked into it and you provided a function void **demo(sol::nested<std::map<std::string, std::vector<std::string>>> src){...}** that I think is what I need, but will have to create a container struct with the variables. I believe I found a typo on the example as well, https://sol2.readthedocs.io/en/latest/api/nested.html . You use g(t) instead of f(t).

**Comment by ThePhD** [comment]
The documentation is wrong, but that code should still work.

**Comment by ThePhD** [comment]
Sorry:

The `sol::table tilesets = tile_table.get<sol::table>("tilesets");` code should work. So should `sol::Table tilesets = tile_table["tilesets"];`.

I would advice against using `sol::nested` because, as you described, you would need to create severely nested structures to properly decompose the entire tileset. It would be less memory intensive and better to just manipulate the series of lua tables present.

**Comment by LedLoaf** [comment]
Okay, great.

So disregarding sol::nested now. Having 'sol::table tilesets = tile_table["tilesets"];' whats the best way to parse each segment? I thought I could maybe grab them as a table like `sol::table v1 = tilesets.get<sol::table>()` which would be nice since if I could just for loop it until the end of tilesets.  What's confusing me is getting ahold of parts that aren't directly labeled like the examples. Is there something with the containers I can do? Where I have a 'tilesets' table and iterate through each object?

**Comment by ThePhD** [comment]
You can us

- `for_each`: https://sol2.readthedocs.io/en/latest/api/table.html#table-for-each
- iterate over it with a ranged for loop, like normal: https://github.com/ThePhD/sol2/blob/develop/examples/source/optional_with_iteration.cpp#L52
- (IF YOU KNOW BEFORE-HAND IT IS A SIZED TABLE WITH INCREMENTAL INTEGER KEYS FROM `1` to `N`, WITH NO GAPS OR `nil`s OR HOLES) `for (int end = my_table.size(), lua_index = 1; start <= end; ++start) { sol::object value = my_table[lua_index]; /* etc... */ }`

**Comment by LedLoaf** [comment]
Yeah, I might be too stupid for this library. 🤦‍♂️😂. It's a bit complex for whatever reason to me. I'm not sure if it's my basic understanding of Lua or what. Let me try one more thing and I'll use it off your example, if I can't understand it at this point I'll just come back in the future.

` lua.script(R"(
		abc = { [0] = 24 }
		def = {
			ghi = {
				bark = 50,
				woof = abc
			},
			{
				bark = 75;
				woof = ABC
			}
		}
	)");`

All I changed was adding the additional array, which is similar to the Tiled Example. You access `ghi` like normal  `int bark1 = def["ghi"]["bark"];` . But the array that has no defined label is what has been difficult to access. I was able to loop the ghi variables, I still can't access the second object and I tried indexing. 

Appreciate your help and time and trying to get me going, I might be stupid for this or need a break (8 hours now).
